src/searchengine/nova3/nova2.py

```diff
@@ -148,47 +193,49 @@ def main(args):
     if current_path not in sys.path:
         sys.path.append(current_path)
 
-    supported_engines = initialize_engines()
+    found_engines = list_engines()
 
     if not args:
         raise SystemExit("./nova2.py [all|engine1[,engine2]*] <category> <keywords>\n"
-                         "available engines: %s" % (','.join(supported_engines)))
+                         "available engines: %s" % (','.join(found_engines)))
 
     elif args[0] == "--capabilities":
+        supported_engines = initialize_engines(found_engines)
         displayCapabilities(supported_engines)
         return
 
     elif len(args) < 3:
         raise SystemExit("./nova2.py [all|engine1[,engine2]*] <category> <keywords>\n"
-                         "available engines: %s" % (','.join(supported_engines)))
+                         "available engines: %s" % (','.join(found_engines)))
+
+    cat = args[1].lower()
+
+    if cat not in CATEGORIES:
+        raise SystemExit(" - ".join(('Invalid category', cat)))
 
     # get only unique engines with set
     engines_list = set(e.lower() for e in args[0].strip().split(','))
 
-    if 'all' in engines_list:
-        engines_list = supported_engines
-    else:
-        # discard un-supported engines
-        engines_list = [engine for engine in engines_list
-                        if engine in supported_engines]
-
     if not engines_list:
         # engine list is empty. Nothing to do here
         return
 
-    cat = args[1].lower()
-
-    if cat not in CATEGORIES:
-        raise SystemExit(" - ".join(('Invalid category', cat)))
+    if 'all' in engines_list:
```

Comment:
Also I believe it would be better (for the further improvements) to get rid from using it this way and require client side to always provide list of used engines explicitly.

Reply:
done. now we force clients to make 2 separate calls: get capabilities, get search results

Reply:
I would remove this branch altogether.

Reply:
hmm. then `nova2.py all all some query` would silently fail
i would keep the `engines = all was removed` error message, to help with debugging

Reply:
If qBittorrent uses "all" then it should be changed as well.

Reply:
`all` is resolved in qBittorrent, not passed to nova2.py

src/gui/search/searchwidget.cpp

```cc
    QStringList plugins;
    if (plugin == u"all")
        plugins = SearchPluginManager::instance()->allPlugins();
    else if ((plugin == u"enabled") || (plugin == u"multi"))
        plugins = SearchPluginManager::instance()->enabledPlugins();
    else
        plugins << plugin;

    qDebug("Search with category: %s", qUtf8Printable(selectedCategory()));

    // Launch search
    auto *searchHandler = SearchPluginManager::instance()->startSearch(pattern, selectedCategory(), plugins);
```

src/base/search/searchpluginmanager.cpp

```cc
QStringList SearchPluginManager::allPlugins() const
{
    return m_plugins.keys();
}

// ...

void SearchPluginManager::update()
{
    QProcess nova;
    nova.setProcessEnvironment(QProcessEnvironment::systemEnvironment());

    const QStringList params
    {
        Utils::ForeignApps::PYTHON_ISOLATE_MODE_FLAG,
        (engineLocation() / Path(u"/nova2.py"_s)).toString(),
        u"--capabilities"_s
    };

// ...

            if (!m_plugins.contains(pluginName))
            {
                m_plugins[pluginName] = plugin.release();
                emit pluginInstalled(pluginName);
            }
            else if (m_plugins[pluginName]->version != plugin->version)
            {
                delete m_plugins.take(pluginName);
                m_plugins[pluginName] = plugin.release();
                emit pluginUpdated(pluginName);
            }

```

Reply:
> `all` is resolved in qBittorrent, not passed to nova2.py

So there's no point to handle "all" in a special way.

Reply:
legacy...
search plugin developers may be used to call `nova2.py all all asdf`
and would be surprised to see nothing

Reply:
This change just needs to be mentioned in Changelog.
IMO, "all" should be considered as regular engine name.

Reply:
sure we want to pollute the qBittorrent changelog with these nitpick changes?
i would prefer `src/searchengine/nova3/Changelog`

i guess its safe to assume that
never in the future there will be a search engine called `all`

Reply:
Now I think we can keep `all` placeholder behavior as it was before to avoid unnecessary code changes. Since qBittorrent itself always uses an explicit list of engines, this should not be a problem.

Reply:
done